File: docs/math/code/binary-exponentiation/luogu-P1226-2.cpp
#include <cstdio>

// --8<-- [start:core]
long long binpow(long long a, long long b, long long p) {
  long long res = 1;
  while (b > 0) {
    if (b & 1) res = res * a % p;
    a = a * a % p;
    b >>= 1;
  }
  return res;
}

// --8<-- [end:core]
int main() {
  int a, b, p;
  scanf("%d%d%d", &a, &b, &p);
  printf("%d^%d mod %d=%lld", a, b, p, binpow(a, b, p));
  return 0;
}

File: math/code/binary-exponentiation/luogu-P1226-2.py
# --8<-- [start:core]
def binpow(a, b, p):
    res = 1
    while b > 0:
        if b & 1:
            res = res * a % p
        a = a * a % p
        b >>= 1
    return res


# --8<-- [end:core]
if __name__ == "__main__":
    a, b, p = map(int, input().split())
    print(f"{a}^{b} mod {p}={binpow(a, b, p)}")
